modules/auxiliary/admin/upnp/soap_portmapping.rb

```diff
@@ -0,0 +1,131 @@
+##
+# encoding: utf-8
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class Metasploit3 < Msf::Auxiliary
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize
+    super(
+      'Name'           => 'UPnP IGD SOAP Port Mapping Utility',
+      'Description'    => 'Manage port mappings on UPnP IGD-capable device using the AddPortMapping and DeletePortMapping SOAP requests',
+      'Author'         =>
+        [
+          'St0rn <fabien[at]anbu-pentest.com>', # initial module
+          'Jon Hart <jon_hart[at]rapid7.com>'   # module cleanup and refactoring
+        ],
+      'License'        => MSF_LICENSE,
+      'References'     => [['URL', 'http://www.upnp-hacks.org/igd.html']],
+      'DefaultAction'  => 'ADD',
+      'Actions'        =>
+        [
+          [ 'ADD',
+            {
+              'Description' => 'Use the AddPortMapping SOAP command to open and forward a port',
+              'SOAP_ACTION' => 'AddPortMapping'
+            }
+          ],
+          [ 'DELETE',
+            {
+              'Description' => 'Use the DeletePortMapping SOAP command to remove a port forwarding',
+              'SOAP_ACTION' => 'DeletePortMapping'
+            }
+          ]
+        ],
+    )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'UPnP control URL', '/' ]),
+        OptAddress.new('INTERNAL_CLIENT', [false, 'Internal client hostname/IP']),
+        OptAddress.new('EXTERNAL_CLIENT', [false, 'External client hostname/IP']),
+        OptEnum.new('PROTOCOL', [true, 'Transport level protocol to map', 'TCP', %w(TCP UDP)]),
+        OptInt.new('INTERNAL_PORT', [false, 'Internal port']),
+        OptInt.new('EXTERNAL_PORT', [true, 'External port']),
+        OptInt.new('LEASE_DURATION', [true, 'Lease time for mapping, in seconds', 3600])
+      ],
+      self.class
+    )
+  end
+
+  def internal_port
+    @internal_port ||= datastore['INTERNAL_PORT']
+  end
+
+  def internal_client
+    @internal_client ||= datastore['INTERNAL_CLIENT']
+  end
+
+  def external_port
+    @external_port ||= datastore['EXTERNAL_PORT']
+  end
+
+  def external_client
+    @external_client ||= datastore['EXTERNAL_CLIENT']
+  end
+
+  def lease_duration
+    @lease_duration ||= datastore['LEASE_DURATION']
+  end
+
+  def protocol
+    @protocol ||= datastore['PROTOCOL']
+  end
+
+  def soap_action
+    @soap_action ||= action.opts['SOAP_ACTION']
+  end
+
+  def run
+    content = "<?xml version=\"1.0\"?>"
```

Comment:
For some time we have been preferring and asking to build XML with REXML/nokogiri instead of by hand, please.